specs/SharedBuffer.md

```diff
@@ -0,0 +1,333 @@
+        if (forFrame)
+        {
+            m_webviewFrame->PostSharedBufferToScript(
+                sharedBuffer.get(), /*isReadOnlyToScript*/TRUE, additionalDataAsJson);
+        }
+        else
+        {
+            m_webView->PostSharedBufferToScript(
+               sharedBuffer.get(), /*isReadOnlyToScript*/TRUE, additionalDataAsJson);
+        }
+        // Explicitly close the one time shared buffer to ensure that the resource is released timely.
+        sharedBuffer->Close();
+        
+```
+## WinRT and .NET
+```c#
+      using (CoreWebView2SharedBuffer sharedBuffer = WebViewEnvironment.CreateSharedBuffer(bufferSize))
+      {
+          // Fill data using access Stream
+          using (Stream stream = sharedBuffer.GetStream())
+          {
+              using (StreamWriter writer = new StreamWriter(stream))
+              {
+                  writer.Write(data);
+              }
+          }
+          string additionalDataAsJson = "{\"readOnly\":true}";
+          if (forFrame)
+          {
+              m_webviewFrame.PostSharedBufferToScript(sharedBuffer, /*isReadOnlyToScript*/true, additionalDataAsJson);
+          }
+          else
+          {
+               m_webview.PostSharedBufferToScript(sharedBuffer, /*isReadOnlyToScript*/true, additionalDataAsJson);
+          }
+      }
+```
+
+# API Details
+## Win32 C++
+```
+interface ICoreWebView2Environment11 : IUnknown {
+  /// Create a shared memory based buffer with the specified size in bytes.
+  /// The buffer can be shared with web contents in WebView by calling
+  /// `PostSharedBufferToScript` on `CoreWebView2` or `CoreWebViewFrame` object.
+  /// Once shared, the same content of the buffer will be accessible from both
+  /// the app process and script in WebView. Modification to the content will be visible
+  /// to all parties that have access to the buffer.
+  /// For 32bit application, the creation will fail with E_INVALIDARG if `size` is larger than 4GB.
+  HRESULT CreateSharedBuffer(
+    [in] UINT64 size,
+    [out, retval] ICoreWebView2SharedBuffer** shared_buffer);
+}
+
+interface ICoreWebView2SharedBuffer : IUnknown {
+  /// The size of the shared buffer in bytes.
+  [propget] HRESULT Size([out, retval] UINT64* value);
+
+  /// The memory address of the shared buffer.
+  [propget] HRESULT Buffer([out, retval] BYTE** value);
+
+  /// Get an IStream object that can be used to access the shared buffer.
+  HRESULT GetStream([out, retval] IStream** value);
+
+  /// The file mapping handle of the shared memory of the buffer.
+  /// Normal app should use `Buffer` or `GetStream` to get memory address
+  /// or IStream object to access the buffer.
+  /// For advanced scenarios, you could duplicate this handle to another application
+  /// process and create a mapping from the duplicated handle in that process to access
+  /// the buffer from that separate process.
+  [propget] HRESULT Handle([out, retval] HANDLE* value);
```

Comment:
No sample showing this. Is the handle owned by the Buffer, or is it duplicated to the caller? Handing out your handle unduplicated could lead to the caller closing it.

Reply:
This is something we need guidance on. As you pointed out, returning un-duped handle could allow app closing it while returning duped one could be a waste of CPU cycle and potential handle leak.

Reply:
I don't think we have a standard, other than documentation.  If you make it a method the semantics are clear - "here is one for you to own."

So you'd change this to be:

```c++
/// Returns a handle to the file mapping object that backs this shared buffer. Your app can
/// use file-mapping APIs to obtain other views or interact with the file mapping, including
/// duplicating it to another process.  The handle that GetHandle returns is owned by the
/// caller; be sure to call CloseHandle on it.
HRESULT GetMappingHandle([out, retval] HANDLE* value);
```

Reply:
Is it acceptable that we document it as "the returned handle is owned by shared buffer object, don't call CloseHandle on it"?

The most likely scenario for using this API is to dup to another process. In this case, returning a duped handle and then ask app to close feels redundant.

But if it is very desirable to return a caller owned handle so that the handle is not accidentally closed by the app, we could definitely change to return a duped handle.

Reply:
Should be dup'ed HANDLE or non-dup'ed HANDLE?

C++: HANDLE MappedFileHandle (non-dup'ed to match COM behavior? Ok to not match C#?)
C#: SafeHandle MappedFileHandle (dup'ed necessary for SafeHandle usage?)
WinRT: Not exposed

C# documentation for MappedFileHandle should note how to use this correctly.

Reply:
As discussed in email thread, always non duped, and SafeHandle for C#.